### PR TITLE
mca/base: add support for long variable values

### PR DESCRIPTION
**Pull request by hjelmn**
Signed-off-by: Nathan Hjelm <hjelmn@lanl.gov>

### CODE REVIEW

File: opal/mca/base/mca_base_var.c [hunk 1]
```diff
@@ -100,7 +101,8 @@ const size_t ompi_var_type_sizes[] = {
     sizeof (char),
     sizeof (char),
     sizeof (bool),
-    sizeof (double)
+    sizeof (double),
+    sizeof (long),
```
Comment (jsquyres) [inline opal/mca/base/mca_base_var.c]: Please remove the trailing comma -- it generates compiler warnings.

Comment (hjelmn) [reply opal/mca/base/mca_base_var.c]: Odd that is complains about that. The trailing comma is allowed by C99 and it is required that the compiler support it. Do you have the warning message handy?

Comment (hjelmn) [reply opal/mca/base/mca_base_var.c]: See C99 6.7.8.

**Comment by hjelmn** [comment]
My use case is I have a variable that needs more than 32-bits but also needs -1.

**Comment by rhc54** [comment]
out of curiosity - why not just go with int64?

**Comment by rhc54** [comment]
I ask because of the indeterminate size of long:
```
Long signed integer type. Capable of containing at least the [−2,147,483,647, +2,147,483,647] range;[3][4] thus, it is at least 32 bits in size.
```
If you are on a system that uses the minimum size, then ```long``` won't meet your described use case.

File: opal/mca/base/mca_base_var.c (hunk 1 above)
Comment (jsquyres) [reply opal/mca/base/mca_base_var.c]: Ah, if this is allowable via C99, then I retract my objection.

**Comment by hjelmn** [comment]
I am fine with int64_t. Long just fit into the current type list. I can update this to add all the fixed width types.

File: opal/mca/base/mca_base_var.c (hunk 1 above)
Comment (hjelmn) [reply opal/mca/base/mca_base_var.c]: I will still look into the message as I am curious what is going on. We use trailing commas all over the place.

**Comment by rhc54** [comment]
fwiw: I avoid leaving trailing commas in the code base because of variation across compilers - some don't bark, others definitely do when using pedantic flags. And yes, I know that isn't strictly what C99 saids - but we all know how closely every compiler adheres to every little detail in that document. Seemed like a trivial thing to simply avoid.

File: opal/mca/base/mca_base_var.c (hunk 1 above)
Comment (jsquyres) [reply opal/mca/base/mca_base_var.c]: Don't spend any time on it.  I am probably remembering something with an old compiler and/or before we enforced C99-specific compiler options.

**Comment by jsquyres** [comment]
If Ralph is seeing those warnings recently, then I would agree with what he said.  I know I have seen these messages before, so I, too, have simply avoided trailing commas.  But I don't have any specific data points about which compiler / platform / etc.

**Comment by hjelmn** [comment]
You know my personal opinion on poorly behaved compilers so I won't repeat it here. I will say I never avoid them but I have also not run into a compiler that c complains about trailing commas.

**Comment by rhc54** [comment]
I wouldn't worry about it - if we see them when building, we can simply submit a PR to remove the comma. I've done similar minor fixes in the past as we all use different compilers, each with their own quirks.

**Comment by hjelmn** [comment]
Added fixed width integer types (both signed and unsigned).

**Comment by ibm-ompi** [comment]
The IBM CI (PGI Compiler) build failed! Please review the log, linked below.

 Gist: https://gist.github.com/24bd28ed4fe46bcc54b44293ef17b6a1

**Comment by hjelmn** [comment]
@jsquyres This look ok to you? I simplified the MCA type -> MPI type code to make it easier to add new types. I doubt we will need to after this PR.

File: ompi/mpi/tool/mpit_common.c [hunk 2]
```diff
@@ -28,53 +28,41 @@ void ompi_mpit_unlock (void)
...
+    [MCA_BASE_VAR_TYPE_UNSIGNED_LONG] = MPI_UNSIGNED_LONG,
+    [MCA_BASE_VAR_TYPE_UNSIGNED_LONG_LONG] = MPI_UNSIGNED_LONG_LONG,
+
+#if SIZEOF_SIZE_T == SIZEOF_UNSIGNED_INT
+    [MCA_BASE_VAR_TYPE_SIZE_T] = MPI_UNSIGNED,
+#elif SIZEOF_SIZE_T == SIZEOF_UNSIGNED_LONG
+    [MCA_BASE_VAR_TYPE_SIZE_T] = MPI_UNSIGNED_LONG,
+#elif SIZEOF_SIZE_T == SIZEOF_LONG_LONG
+    [MCA_BASE_VAR_TYPE_SIZE_T] = MPI_UNSIGNED_LONG_LONG,
+#else
```
Comment (rhc54) [inline ompi/mpi/tool/mpit_common.c]: @hjelmn do you want to add the 64-bit and 32-bit fixed size options here?

Comment (rhc54) [reply ompi/mpi/tool/mpit_common.c]: Other than that possible change, looks good. Thanks!

**Pull request merged**